SeQuant/core/export/python_einsum.hpp

```diff
@@ -0,0 +1,636 @@
+
+#include <cctype>
+#include <cstdlib>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace sequant {
+
+/// Backend selection for Python einsum generator
+enum class PythonEinsumBackend {
+  NumPy,   ///< Generate code using numpy.einsum
+  PyTorch  ///< Generate code using torch.einsum
+};
+
+/// Context class for the PythonEinsumGenerator
+class PythonEinsumGeneratorContext : public ExportContext {
+ public:
+  using ShapeMap = std::map<IndexSpace, std::string>;
+
+  PythonEinsumGeneratorContext(
+      PythonEinsumBackend backend = PythonEinsumBackend::NumPy)
+      : m_backend(backend) {}
+  ~PythonEinsumGeneratorContext() = default;
+  PythonEinsumGeneratorContext(
+      ShapeMap index_shapes,
+      PythonEinsumBackend backend = PythonEinsumBackend::NumPy)
+      : m_index_shapes(std::move(index_shapes)), m_backend(backend) {}
+
+  /// Get the dimension/shape for a given index space
+  std::string get_shape(const IndexSpace &space) const {
+    auto it = m_index_shapes.find(space);
+    if (it != m_index_shapes.end()) {
+      return it->second;
+    }
+    // Default fallback - use a simple naming scheme
+    return "dim_" + toUtf8(space.base_key());
+  }
+
+  /// Get the shape tuple for a tensor
+  std::string get_shape_tuple(const Tensor &tensor) const {
+    std::string shape = "(";
+    bool first = true;
+    for (const Index &idx : tensor.const_indices()) {
+      if (!first) shape += ", ";
+      shape += get_shape(idx.space());
+      first = false;
+    }
+    shape += ")";
+    return shape;
+  }
+
+  /// Set the dimension/shape for a given index space
+  void set_shape(const IndexSpace &space, std::string shape) {
+    m_index_shapes[space] = std::move(shape);
+  }
+
+  /// Get the backend being used
+  PythonEinsumBackend backend() const { return m_backend; }
+
+  /// Set the backend
+  void set_backend(PythonEinsumBackend backend) { m_backend = backend; }
+
+  /// Get the module prefix (np. or torch.)
+  std::string module_prefix() const {
+    return m_backend == PythonEinsumBackend::NumPy ? "np." : "torch.";
+  }
+
+ protected:
+  ShapeMap m_index_shapes;
+  PythonEinsumBackend m_backend;
+};
+
+/// Generator for producing Python code using numpy.einsum or torch.einsum
+/// - NumPy: https://numpy.org/doc/stable/reference/generated/numpy.einsum.html
+/// - PyTorch: https://pytorch.org/docs/stable/generated/torch.einsum.html
+template <typename Context = PythonEinsumGeneratorContext>
+class PythonEinsumGenerator : public Generator<Context> {
+ public:
+  PythonEinsumGenerator() = default;
+  ~PythonEinsumGenerator() = default;
+
+  std::string get_format_name() const override { return "Python (einsum)"; }
+
+  bool supports_named_sections() const override { return true; }
+
+  bool requires_named_sections() const override { return false; }
+
+  DeclarationScope index_declaration_scope() const override {
+    return DeclarationScope::Global;
```

Comment:
I believe `Section` would be more appropriate. Index declarations can be (ab)used to add parameters representing index dimensions to function (aka: "named section") signatures. Otherwise, the variables holding the index dimensions would have to be global.

Reply:
If you want to make the generated code rely on global variables (as it currently does), whatever this function returns is insignificant. My suggestion was to make use of index declarations in order to generate functions that take the index dimensions as parameters:
```python
def bla(n_occ, n_virt):
    pass
```
for which we would need `DeclarationScope::Section`